File: programmers/level2/쿼드압축 후 개수 세기.py
def solution(arr):
    def dfs(posy, posx, line_len) :
        temp = arr[posy][posx]
        check = False
        for i in range(posy, posy+line_len):
            for j in range(posx, posx+line_len):
                if arr[i][j] != temp:
                    check = True
                    dfs(posy, posx, line_len//2)
                    dfs(posy+line_len//2, posx, line_len//2)
                    dfs(posy, posx+line_len//2, line_len//2)
                    dfs(posy+line_len//2, posx+line_len//2, line_len//2)
                    break
            if check: break
        if not check:
            ans[temp] += 1
    ans = [0,0]
    line_len = len(arr)
    posy, posx = 0, 0
    dfs(posy, posx, line_len)
    return ans

File: programmers/level2/쿼드압축 후 개수 세기.java
class Solution {
    static int[] answer = new int[2];
    int[][] arr;
    
    public int[] solution(int[][] arr) {
        this.arr = arr;
        int width = arr.length;
        
        dfs(0,0,width);
        return answer;
    }
    public void dfs(int y, int x, int length) {
        int n = arr[y][x];
        boolean check = false;
        for(int nowY=y; nowY<y+length; nowY++) {
            for(int nowX=x; nowX<x+length; nowX++) {
                if(arr[nowY][nowX] != n) {
                    check = true;
                    dfs(y, x, length/2);
                    dfs(y+length/2, x, length/2);
                    dfs(y, x+length/2, length/2);
                    dfs(y+length/2, x+length/2, length/2);
                    break;
                }
            }
            if(check) break;
        }
        if(!check) {
            answer[n]++;
        }
    }
}